lib/auth/agent_version_report.go

```diff
@@ -0,0 +1,162 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	autoupdatev1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/autoupdate/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/autoupdate"
+	"github.com/gravitational/teleport/lib/inventory"
+)
+
+type instanceReport struct {
+	data      map[string]instanceGroupReport
+	timestamp time.Time
+}
+
+func (ir instanceReport) collectInstance(handle inventory.UpstreamHandle) {
+	// If the instance is being soft-reloaded or shut down, we ignore it.
+	if goodbye := handle.Goodbye(); goodbye != nil && (goodbye.GetSoftReload() || goodbye.GetDeleteResources()) {
+		return
+	}
+
+	// We skip servers that joined less than a minute ago as they might have been
+	// connected to another auth instance a few seconds ago, which would lead to double-counting.
+	if handle.RegistrationTime().After(ir.timestamp.Add(-time.Minute)) {
+		return
+	}
+	// We skip auth and proxy instances because we don't update them.
+	if handle.HasService(types.RoleAuth) || handle.HasService(types.RoleProxy) {
```

Comment:
nit: we could augment `inventory.UpstreamHandle` with a `HasControlPlaneService` that calls `(types.SystemRole).IsControlPlane()` on the roles it has.